Add high availability (#113)
ConsenSys/doc.tessera: docs/HowTo/Configure/High-availability.md

---
description: Tessera deployment for high availability
---

# High availability

Tessera supports deploying more than one instance sharing the same database.

> By placing the instances behind a load balancer, downtime can be limited during maintenance operations.

Reviewer: ```suggestion
By placing the instances behind a load balancer, downtime can be limited during maintenance operations, achieving high availability (HA).
```

> ![Tessera-HA](../../images/tessera/Tessera-HA.png)

Reviewer: Nitpick, but can you lessen the gray space at the top of the diagram and make "Balancer" and "Client" lowercase?

## Servers

Tessera exposes multiple interfaces for connectivity which can be configured in the `serverConfigs` in the
[configuration file](../../Reference/SampleConfiguration.md).

> To enable High availability for the node, set the interfaces to use the load balancer's address for their `serverAddress`,
> and their own URL or IP for the `bindingAddress`, as shown in the following example.

Reviewer: ```suggestion
To enable high availability for the node, set each interface to use the load balancer's address for its `serverAddress`,
and its own URL or IP for the `bindingAddress`, as shown in the following example.
```

> ```bash

Reviewer: ```suggestion
!!! example "Server configuration"

    ```bash
```

    ```
     events { }

     http {
         upstream tessera_tp_9080 {
             server tessera_1:9080 max_fails=3 fail_timeout=5s;
             server tessera_2:9080 max_fails=3 fail_timeout=5s;
         }

         upstream tessera_q2t_9101 {
             server tessera_1:9101 max_fails=3 fail_timeout=5s;
             server tessera_2:9101 max_fails=3 fail_timeout=5s;
         }

          upstream tessera_p2p_9000 {
              server tessera_1:9000 max_fails=3 fail_timeout=5s;
              server tessera_2:9000 max_fails=3 fail_timeout=5s;
          }

         server {
             listen 9080;

             location / {
                 proxy_pass http://tessera_tp_9080;
                 health_check port=9000 uri=/upcheck;
             }
         }

         server {
             listen 9101;

             location / {
                 proxy_pass http://tessera_q2t_9101;
                 health_check port=9000 uri=/upcheck;
             }
         }

        server {
            listen 9000;

            location / {
                proxy_pass http://tessera_p2p_9000;
                health_check port=9000 uri=/upcheck;
            }
        }
     }
    ```

> The configuration defines two upstreams: `tessera_tp_9080` and `tessera_q2t_9101`.
> 
> Both upstreams define health checks: `max_fails=3 fail_timeout=5s`
> 
> The [health checks](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/) help Nginx balance

Reviewer: ```suggestion
The configuration defines two upstreams, `tessera_tp_9080` and `tessera_q2t_9101`, both of which define [health checks](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/), `max_fails=3 fail_timeout=5s`.

The health checks help Nginx balance
```

traffic among upstream servers.

## Database

> The last piece to configure in High Availability is the [database](./Database.md).

Reviewer: ```suggestion
The last piece to configure in high availability is the [database](./Database.md).
```
